[FSDP] Consolidate FSDP state_dict offload_to_cpu settings

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86912
* #86243
* __->__ #86211

Consolidate FSDP state_dict offload_to_cpu settings. All state_dict_types now
have offload_to_cpu options.

Differential Revision: [D40065969](https://our.internmc.facebook.com/intern/diff/D40065969/)

fegin committed:
[FSDP] Consolidate FSDP state_dict offload_to_cpu settings

Consolidate FSDP state_dict offload_to_cpu settings. All state_dict_types now
have offload_to_cpu options.

Differential Revision: [D40065969](https://our.internmc.facebook.com/intern/diff/D40065969/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM, thanks! Do we need to update the appropriate documentation anywhere?

awgu commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -180,8 +185,16 @@ def _get_state_dict_mgr(
+            )
         else:
-            config = None
+            raise ValueError("Unspported state_dict_type")
```
nit:
```suggestion
            raise ValueError("Unsupported state_dict_type")
```

awgu commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -13,7 +13,12 @@
+    ShardedStateDictConfig,
+)
 from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
 from torch.distributed.fsdp import (
```
nit: Possibly sort imports to fix the lint?

fegin commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -73,8 +78,8 @@
-_UNFLATTENED_STATE_DICT_IMPLS = ["state_dict", "sharded_state_dict"]
-_FLATTENED_STATE_DICT_IMPLS = ["local_state_dict"]
+_UNFLATTENED_STATE_DICT_IMPLS = ["state_dict"]
+_FLATTENED_STATE_DICT_IMPLS = []
```
oops, accidentally commit this change. Revert it.

fegin committed:
Update on "[FSDP] Consolidate FSDP state_dict offload_to_cpu settings"


Consolidate FSDP state_dict offload_to_cpu settings. All state_dict_types now
have offload_to_cpu options.

Differential Revision: [D40065969](https://our.internmc.facebook.com/intern/diff/D40065969/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Consolidate FSDP state_dict offload_to_cpu settings"


Consolidate FSDP state_dict offload_to_cpu settings. All state_dict_types now
have offload_to_cpu options.

Differential Revision: [D40065969](https://our.internmc.facebook.com/intern/diff/D40065969/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Consolidate FSDP state_dict offload_to_cpu settings"


Consolidate FSDP state_dict offload_to_cpu settings. All state_dict_types now
have offload_to_cpu options.

Differential Revision: [D40065969](https://our.internmc.facebook.com/intern/diff/D40065969/)

[ghstack-poisoned]

fegin commented:
@pytorchbot merge
